### ISSUE TITLE
MODULE FAILURE

### ISSUE THREAD

**Issue by Linuxfabrik** [ansible-collections/community.digitalocean#199]
##### SUMMARY
Deploying a droplet fails with "MODULE FAILURE: AttributeError: 'NoneType' object has no attribute 'get'"

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME
community.digitalocean.digital_ocean_droplet

##### ANSIBLE VERSION
```
ansible 2.9.27
  config file = /opt/ansible/ansible.cfg
  configured module search path = ['/root/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python3.10/site-packages/ansible
  executable location = /usr/bin/ansible
  python version = 3.10.0 (default, Oct  4 2021, 00:00:00) [GCC 11.2.1 20210728 (Red Hat 11.2.1-1)]
```

##### COLLECTION VERSION
```
1.13.0
```

##### CONFIGURATION
```
ANSIBLE_FORCE_COLOR(/opt/ansible/ansible.cfg) = True
ANSIBLE_NOCOWS(/opt/ansible/ansible.cfg) = True
ANSIBLE_PIPELINING(/opt/ansible/ansible.cfg) = True
CACHE_PLUGIN(/opt/ansible/ansible.cfg) = jsonfile
CACHE_PLUGIN_CONNECTION(/opt/ansible/ansible.cfg) = ~/.ansible_cache
CACHE_PLUGIN_TIMEOUT(/opt/ansible/ansible.cfg) = 86400
DEFAULT_CALLBACK_WHITELIST(/opt/ansible/ansible.cfg) = ['profile_tasks']
DEFAULT_FORKS(/opt/ansible/ansible.cfg) = 30
DEFAULT_GATHERING(/opt/ansible/ansible.cfg) = smart
DEFAULT_HOST_LIST(/opt/ansible/ansible.cfg) = ['/opt/ansible/hosts']
DEFAULT_LOG_PATH(/opt/ansible/ansible.cfg) = /root/.ansible/ansible.log
DEFAULT_MANAGED_STR(/opt/ansible/ansible.cfg) = This file is managed by Ansible - do not edit
DEFAULT_ROLES_PATH(/opt/ansible/ansible.cfg) = ['/opt/ansible/roles']
DEFAULT_STDOUT_CALLBACK(/opt/ansible/ansible.cfg) = yaml
DEFAULT_TIMEOUT(/opt/ansible/ansible.cfg) = 5
HOST_KEY_CHECKING(/opt/ansible/ansible.cfg) = False
INVENTORY_IGNORE_EXTS(/opt/ansible/ansible.cfg) = ['~', '.orig', '.bak', '.ini', '.cfg', '.retry', '.pyc', '.pyo', '.csv']
RETRY_FILES_ENABLED(/opt/ansible/ansible.cfg) = False
```

##### OS / ENVIRONMENT
Linux fedora35ws 5.15.6-200.fc35.x86_64 #1 SMP Wed Dec 1 13:41:10 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
Python 3.10.0

##### STEPS TO REPRODUCE
```yaml
- name: first play
  hosts: localhost
  connection: local

  tasks:
    - name: 'Create a VM at DigitalOcean'
      community.digitalocean.digital_ocean_droplet:
        name: 'vmname'
        oauth_token: 'token'        
        ssh_keys:
          - '12:34:56'
        size: 's-1vcpu-1gb'
        image: 'rockylinux-8-x64'
        region: 'fra1'
      delegate_to: 'localhost'
      register: my_droplet

    - debug:
        msg: "ID is {{ my_droplet.data.droplet.id }}, IP is {{ my_droplet.data.ip_address }}"
```

```bash
ansible-playbook playbooks/cloud.yml -vvv
```

##### EXPECTED RESULTS
Printing IP address of new Droplet.

##### ACTUAL RESULTS
Droplet is created, but when trying to print IP address, the Ansible plugin fails.

```
Using module file /root/.ansible/collections/ansible_collections/community/digitalocean/plugins/modules/digital_ocean_droplet.py
Pipelining is enabled.
<localhost> ESTABLISH LOCAL CONNECTION FOR USER: root
<localhost> EXEC /bin/sh -c '/usr/bin/python3 && sleep 0'
The full traceback is:
Traceback (most recent call last):
  File "<stdin>", line 102, in <module>
  File "<stdin>", line 94, in _ansiballz_main
  File "<stdin>", line 40, in invoke_module
  File "/usr/lib64/python3.10/runpy.py", line 209, in run_module
    return _run_module_code(code, init_globals, run_name, mod_spec)
  File "/usr/lib64/python3.10/runpy.py", line 96, in _run_module_code
    _run_code(code, mod_globals, init_globals,
  File "/usr/lib64/python3.10/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "/tmp/ansible_community.digitalocean.digital_ocean_droplet_payload_lxcj3rti/ansible_community.digitalocean.digital_ocean_droplet_payload.zip/ansible_collections/community/digitalocean/plugins/modules/digital_ocean_droplet.py", line 894, in <module>
  File "/tmp/ansible_community.digitalocean.digital_ocean_droplet_payload_lxcj3rti/ansible_community.digitalocean.digital_ocean_droplet_payload.zip/ansible_collections/community/digitalocean/plugins/modules/digital_ocean_droplet.py", line 890, in main
  File "/tmp/ansible_community.digitalocean.digital_ocean_droplet_payload_lxcj3rti/ansible_community.digitalocean.digital_ocean_droplet_payload.zip/ansible_collections/community/digitalocean/plugins/modules/digital_ocean_droplet.py", line 843, in core
  File "/tmp/ansible_community.digitalocean.digital_ocean_droplet_payload_lxcj3rti/ansible_community.digitalocean.digital_ocean_droplet_payload.zip/ansible_collections/community/digitalocean/plugins/modules/digital_ocean_droplet.py", line 760, in create
AttributeError: 'NoneType' object has no attribute 'get'
fatal: [localhost -> localhost]: FAILED! => changed=false 
  module_stderr: |-
    Traceback (most recent call last):
      File "<stdin>", line 102, in <module>
      File "<stdin>", line 94, in _ansiballz_main
      File "<stdin>", line 40, in invoke_module
      File "/usr/lib64/python3.10/runpy.py", line 209, in run_module
        return _run_module_code(code, init_globals, run_name, mod_spec)
      File "/usr/lib64/python3.10/runpy.py", line 96, in _run_module_code
        _run_code(code, mod_globals, init_globals,
      File "/usr/lib64/python3.10/runpy.py", line 86, in _run_code
        exec(code, run_globals)
      File "/tmp/ansible_community.digitalocean.digital_ocean_droplet_payload_lxcj3rti/ansible_community.digitalocean.digital_ocean_droplet_payload.zip/ansible_collections/community/digitalocean/plugins/modules/digital_ocean_droplet.py", line 894, in <module>
      File "/tmp/ansible_community.digitalocean.digital_ocean_droplet_payload_lxcj3rti/ansible_community.digitalocean.digital_ocean_droplet_payload.zip/ansible_collections/community/digitalocean/plugins/modules/digital_ocean_droplet.py", line 890, in main
      File "/tmp/ansible_community.digitalocean.digital_ocean_droplet_payload_lxcj3rti/ansible_community.digitalocean.digital_ocean_droplet_payload.zip/ansible_collections/community/digitalocean/plugins/modules/digital_ocean_droplet.py", line 843, in core
      File "/tmp/ansible_community.digitalocean.digital_ocean_droplet_payload_lxcj3rti/ansible_community.digitalocean.digital_ocean_droplet_payload.zip/ansible_collections/community/digitalocean/plugins/modules/digital_ocean_droplet.py", line 760, in create
    AttributeError: 'NoneType' object has no attribute 'get'
  module_stdout: ''
  msg: |-
    MODULE FAILURE
    See stdout/stderr for the exact error
  rc: 1

```

**Comment by mamercad** [comment]
Looks like [the API](https://docs.digitalocean.com/reference/api/api-reference/#operation/create_droplet) might have changed on us, looks like when creating a Droplet, this is what we'll get back (doesn't have an `ip_address` key):

```json
"droplet": {
"id": 3164444,
"name": "example.com",
"memory": 1024,
"vcpus": 1,
"disk": 25,
"locked": false,
"status": "new",
"kernel": null,
"created_at": "2020-07-21T18:37:44Z",
"features": [],
"backup_ids": [ ],
"next_backup_window": null,
"snapshot_ids": [ ],
"image": {},
"volume_ids": [ ],
"size": {},
"size_slug": "s-1vcpu-1gb",
"networks": {},
"region": {},
"tags": []
}
```

**Comment by mamercad** [comment]
Looks like we'll have to update docs and examples to use the "networks" key:

```
      networks:
        v4:
        - gateway: 161.35.96.1
          ip_address: 161.35.106.87
          netmask: 255.255.240.0
          type: public
        - gateway: ''
          ip_address: 10.116.0.2
          netmask: 255.255.240.0
          type: private
        v6: []
```

**Comment by francoposa** [comment]
I am also having this issue, using the updated examples. I can offer a little more detail:

- droplet is always created, that's no issue.
- `wait: no` "avoids" this issue, seems to skip the part of the code that hits this
- when waiting, `wait_timeout` is not respected. Pushing it up to `1000` does not make the module wait any longer before failing out

playbook:
```yaml
---
- hosts: localhost
  tasks:
    - name: create a new droplet in project "demo"
      community.digitalocean.digital_ocean_droplet:
        state: active
        oauth_token: "{{ lookup('ansible.builtin.env', 'DO_API_TOKEN') }}"
        name: demo-sfo3-01
        size: s-1vcpu-1gb
        region: sfo3
        image: centos-stream-8-x64
        wait_timeout: 1000
        ssh_keys:
          - '12:34:56'
        project: demo
```

On latest version of ansible package on pypi, `4.10.0`, with whichever version of the digitalocean plugin comes with that.

Everything is default

```shell
% ansible --version
ansible [core 2.11.7] 
  config file = None
  configured module search path = ['/Users/franco/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/franco/.virtualenvs/ops/lib/python3.9/site-packages/ansible
  ansible collection location = /Users/franco/.ansible/collections:/usr/share/ansible/collections
  executable location = /Users/franco/.virtualenvs/ops/bin/ansible
  python version = 3.9.7 (default, Sep  3 2021, 18:14:31) [Clang 12.0.5 (clang-1205.0.22.9)]
  jinja version = 3.0.3
  libyaml = True
```


Full verbose error output:

```shell
TASK [create a new droplet in project "demo"] ******************************************************************************************************
task path: /Users/franco/repos/learn-infra-ops/cloud-infra/ansible/inventory/digitalocean/create-digitalocean-inventory.yaml:4
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: franco
<127.0.0.1> EXEC /bin/sh -c 'echo ~franco && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /Users/franco/.ansible/tmp `"&& mkdir "` echo /Users/franco/.ansible/tmp/ansible-tmp-1641242265.899792-10495-226892805387363 `" && echo ansible-tmp-1641242265.899792-10495-226892805387363="` echo /Users/franco/.ansible/tmp/ansible-tmp-1641242265.899792-10495-226892805387363 `" ) && sleep 0'
Using module file /Users/franco/.virtualenvs/ops/lib/python3.9/site-packages/ansible_collections/community/digitalocean/plugins/modules/digital_ocean_droplet.py
<127.0.0.1> PUT /Users/franco/.ansible/tmp/ansible-local-1046086uj6oh7/tmpk6paq1ad TO /Users/franco/.ansible/tmp/ansible-tmp-1641242265.899792-10495-226892805387363/AnsiballZ_digital_ocean_droplet.py
<127.0.0.1> EXEC /bin/sh -c 'chmod u+x /Users/franco/.ansible/tmp/ansible-tmp-1641242265.899792-10495-226892805387363/ /Users/franco/.ansible/tmp/ansible-tmp-1641242265.899792-10495-226892805387363/AnsiballZ_digital_ocean_droplet.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '/Users/franco/.virtualenvs/ops/bin/python /Users/franco/.ansible/tmp/ansible-tmp-1641242265.899792-10495-226892805387363/AnsiballZ_digital_ocean_droplet.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c 'rm -f -r /Users/franco/.ansible/tmp/ansible-tmp-1641242265.899792-10495-226892805387363/ > /dev/null 2>&1 && sleep 0'
The full traceback is:
Traceback (most recent call last):
  File "/Users/franco/.ansible/tmp/ansible-tmp-1641242265.899792-10495-226892805387363/AnsiballZ_digital_ocean_droplet.py", line 100, in <module>
    _ansiballz_main()
  File "/Users/franco/.ansible/tmp/ansible-tmp-1641242265.899792-10495-226892805387363/AnsiballZ_digital_ocean_droplet.py", line 92, in _ansiballz_main
    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)
  File "/Users/franco/.ansible/tmp/ansible-tmp-1641242265.899792-10495-226892805387363/AnsiballZ_digital_ocean_droplet.py", line 40, in invoke_module
    runpy.run_module(mod_name='ansible_collections.community.digitalocean.plugins.modules.digital_ocean_droplet', init_globals=dict(_module_fqn='ansible_collections.community.digitalocean.plugins.modules.digital_ocean_droplet', _modlib_path=modlib_path),
  File "/Users/franco/.pyenv/versions/3.9.7/lib/python3.9/runpy.py", line 210, in run_module
    return _run_module_code(code, init_globals, run_name, mod_spec)
  File "/Users/franco/.pyenv/versions/3.9.7/lib/python3.9/runpy.py", line 97, in _run_module_code
    _run_code(code, mod_globals, init_globals,
  File "/Users/franco/.pyenv/versions/3.9.7/lib/python3.9/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/var/folders/g5/rfzgnx0n61g2d6c11dj6vwbc0000gn/T/ansible_community.digitalocean.digital_ocean_droplet_payload_dv77bi6u/ansible_community.digitalocean.digital_ocean_droplet_payload.zip/ansible_collections/community/digitalocean/plugins/modules/digital_ocean_droplet.py", line 894, in <module>
  File "/var/folders/g5/rfzgnx0n61g2d6c11dj6vwbc0000gn/T/ansible_community.digitalocean.digital_ocean_droplet_payload_dv77bi6u/ansible_community.digitalocean.digital_ocean_droplet_payload.zip/ansible_collections/community/digitalocean/plugins/modules/digital_ocean_droplet.py", line 890, in main
  File "/var/folders/g5/rfzgnx0n61g2d6c11dj6vwbc0000gn/T/ansible_community.digitalocean.digital_ocean_droplet_payload_dv77bi6u/ansible_community.digitalocean.digital_ocean_droplet_payload.zip/ansible_collections/community/digitalocean/plugins/modules/digital_ocean_droplet.py", line 843, in core
  File "/var/folders/g5/rfzgnx0n61g2d6c11dj6vwbc0000gn/T/ansible_community.digitalocean.digital_ocean_droplet_payload_dv77bi6u/ansible_community.digitalocean.digital_ocean_droplet_payload.zip/ansible_collections/community/digitalocean/plugins/modules/digital_ocean_droplet.py", line 760, in create
AttributeError: 'NoneType' object has no attribute 'get'
fatal: [localhost]: FAILED! => {
    "changed": false,
    "module_stderr": "Traceback (most recent call last):\n  File \"/Users/franco/.ansible/tmp/ansible-tmp-1641242265.899792-10495-226892805387363/AnsiballZ_digital_ocean_droplet.py\", line 100, in <module>\n    _ansiballz_main()\n  File \"/Users/franco/.ansible/tmp/ansible-tmp-1641242265.899792-10495-226892805387363/AnsiballZ_digital_ocean_droplet.py\", line 92, in _ansiballz_main\n    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)\n  File \"/Users/franco/.ansible/tmp/ansible-tmp-1641242265.899792-10495-226892805387363/AnsiballZ_digital_ocean_droplet.py\", line 40, in invoke_module\n    runpy.run_module(mod_name='ansible_collections.community.digitalocean.plugins.modules.digital_ocean_droplet', init_globals=dict(_module_fqn='ansible_collections.community.digitalocean.plugins.modules.digital_ocean_droplet', _modlib_path=modlib_path),\n  File \"/Users/franco/.pyenv/versions/3.9.7/lib/python3.9/runpy.py\", line 210, in run_module\n    return _run_module_code(code, init_globals, run_name, mod_spec)\n  File \"/Users/franco/.pyenv/versions/3.9.7/lib/python3.9/runpy.py\", line 97, in _run_module_code\n    _run_code(code, mod_globals, init_globals,\n  File \"/Users/franco/.pyenv/versions/3.9.7/lib/python3.9/runpy.py\", line 87, in _run_code\n    exec(code, run_globals)\n  File \"/var/folders/g5/rfzgnx0n61g2d6c11dj6vwbc0000gn/T/ansible_community.digitalocean.digital_ocean_droplet_payload_dv77bi6u/ansible_community.digitalocean.digital_ocean_droplet_payload.zip/ansible_collections/community/digitalocean/plugins/modules/digital_ocean_droplet.py\", line 894, in <module>\n  File \"/var/folders/g5/rfzgnx0n61g2d6c11dj6vwbc0000gn/T/ansible_community.digitalocean.digital_ocean_droplet_payload_dv77bi6u/ansible_community.digitalocean.digital_ocean_droplet_payload.zip/ansible_collections/community/digitalocean/plugins/modules/digital_ocean_droplet.py\", line 890, in main\n  File \"/var/folders/g5/rfzgnx0n61g2d6c11dj6vwbc0000gn/T/ansible_community.digitalocean.digital_ocean_droplet_payload_dv77bi6u/ansible_community.digitalocean.digital_ocean_droplet_payload.zip/ansible_collections/community/digitalocean/plugins/modules/digital_ocean_droplet.py\", line 843, in core\n  File \"/var/folders/g5/rfzgnx0n61g2d6c11dj6vwbc0000gn/T/ansible_community.digitalocean.digital_ocean_droplet_payload_dv77bi6u/ansible_community.digitalocean.digital_ocean_droplet_payload.zip/ansible_collections/community/digitalocean/plugins/modules/digital_ocean_droplet.py\", line 760, in create\nAttributeError: 'NoneType' object has no attribute 'get'\n",
    "module_stdout": "",
    "msg": "MODULE FAILURE\nSee stdout/stderr for the exact error",
    "rc": 1
}

PLAY RECAP *****************************************************************************************************************************************
localhost                  : ok=1    changed=0    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0
```

**Comment by francoposa** [comment]
Confirmed that this is resolved in digitalocean collection version 1.14, which is the one used by Ansible 5.1

**Comment by mamercad** [comment]
Going to close this since it's resolved in the current release.
